### PR TITLE
Court nav case assignment to court

### PR DESCRIPTION
**Pull request by khurshedk**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRL-5787
https://tools.hmcts.net/jira/browse/PRL-5619

### Change description ###

EPIMS id passed by Cournav - Assign the case to court
EMIMS id not available, check the LD flag, if flag is on - assign the case to Swansea if flag is on
EMIMS id not available, check the LD flag, if flag is off - no court is assigned

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java [hunk 1]
```diff
@@ -217,17 +227,57 @@ public CaseData mapCourtNavData(CourtNavFl401 courtNavCaseData) throws NotFoundE
...
+        caseData = populateCourtDetailsForCourtNaveCase(authorization, caseData,
+                                                        courtNavCaseData.getMetaData().getCourtSpecialRequirements());
         caseData = caseData.setDateSubmittedDate();
 
         return caseData;
 
     }
 
+    private CaseData populateCourtDetailsForCourtNaveCase(String authorization, CaseData caseData,
+                                                          String epmsId) throws NotFoundException {
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:239]: its called EPIMS id please.

Comment (AlokDatta) [reply src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:239]: Do you wanna run a null check on metadata before calling this method?

File: src/main/java/uk/gov/hmcts/reform/prl/services/TestingSupportService.java [hunk 2]
```diff
@@ -250,8 +250,8 @@ private Map<String, Object> updateCaseDetailsForCourtNav(CaseDetails initialCase
                                                              CourtNavFl401 dummyCaseDetails) throws NotFoundException {
         Map<String, Object> caseDataUpdated = new HashMap<>();
         if (dummyCaseDetails != null) {
-            CaseData fl401CourtNav = fl401ApplicationMapper.mapCourtNavData(dummyCaseDetails);
             String systemAuthorisation = systemUserService.getSysUserToken();
+            CaseData fl401CourtNav = fl401ApplicationMapper.mapCourtNavData(dummyCaseDetails,systemAuthorisation);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/TestingSupportService.java:254]: format please.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/courtnav/CourtNavCaseController.java [hunk 3]
```diff
@@ -65,7 +65,7 @@ public ResponseEntity<Object> createCase(
 
         if (Boolean.TRUE.equals(authorisationService.authoriseUser(authorisation)) && Boolean.TRUE.equals(
             authorisationService.authoriseService(serviceAuthorization))) {
-            CaseData caseData = fl401ApplicationMapper.mapCourtNavData(inputData);
+            CaseData caseData = fl401ApplicationMapper.mapCourtNavData(inputData,authorisation);
```
[the remark this answers is not in the document]
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/controllers/courtnav/CourtNavCaseController.java:68]: formatting done

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java [hunk 4]
```diff
@@ -88,9 +94,13 @@
 public class FL401ApplicationMapper {
 
     private final CourtFinderService courtFinderService;
+    private final LaunchDarklyClient launchDarklyClient;
+    public static final String FL401_DEFAULT_BASE_LOCATION_ID = "234946";
+    private final LocationRefDataService locationRefDataService;
     private Court court = null;
+    public static final String COURTNAV_SWANSEA_COURT_MAPPING = "courtnav-swansea-court-mapping";
```
[the remark this answers is not in the document]
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:101]: moved this to PrlLaunchDarklyFlagConstants

File: src/main/java/uk/gov/hmcts/reform/prl/services/TestingSupportService.java (hunk 2 above)
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/services/TestingSupportService.java:254]: formatting done

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java (hunk 1 above)
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:239]: I think null check is not required here as metadata is already getting used before calling this method.

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java [hunk 5]
```diff
@@ -217,17 +227,58 @@ public CaseData mapCourtNavData(CourtNavFl401 courtNavCaseData) throws NotFoundE
...
-
+        caseData = populateCourtDetailsForCourtNaveCase(authorization, caseData,
+                                                        courtNavCaseData.getMetaData().getCourtSpecialRequirements());
         caseData = caseData.setDateSubmittedDate();
 
         return caseData;
 
     }
 
+    private CaseData populateCourtDetailsForCourtNaveCase(String authorization, CaseData caseData,
```
Comment (yogendraU) [inline src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:238]: please remove e from court nav

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java [hunk 6]
```diff
@@ -217,17 +227,58 @@ public CaseData mapCourtNavData(CourtNavFl401 courtNavCaseData) throws NotFoundE
...
+                                                          String epimsId) throws NotFoundException {
+        Optional<CourtVenue> courtVenue = Optional.empty();
+        //1. get court details from provided epimsId request
+        if (!StringUtils.isEmpty(epimsId)) {
+            courtVenue = getCourtVenue(authorization, epimsId);
+        }
+        //2. if not found check launch-darkly flag and populate default Swansea court Id.
+        if (launchDarklyClient.isFeatureEnabled(PrlLaunchDarklyFlagConstants.COURTNAV_SWANSEA_COURT_MAPPING)
+            && courtVenue.isEmpty()) {
+            epimsId = COURTNAV_DUMMY_BASE_LOCATION_ID;
```
Comment (yogendraU) [inline src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:248]: I think we should give epimsId of swansea?
@AlokDatta , what do you say.

**Review by yogendraU — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java (hunk 5 above)
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:238]: corrected as per comment

File: src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java (hunk 6 above)
Comment (khurshedk) [reply src/main/java/uk/gov/hmcts/reform/prl/courtnav/mappers/FL401ApplicationMapper.java:248]: @yogendraU this COURTNAV_DUMMY_BASE_LOCATION_ID contains epims id of swansea court.
cc @AlokDatta

**Review by AlokDatta — APPROVED** [review]

**Review by AlokDatta — APPROVED** [review]

**Pull request merged**